src/viewport-observer.js

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {isIframed} from './dom';
+
+/**
+ * Returns an IntersectionObserver tracking the Viewport.
+ * Only use this if x-origin iframe rootMargin support is considered nice-to-have,
+ * since if not supported this InOb will fallback to one without rootMargin.
+ *
+ * - If iframed: rootMargin is ignored unless natively supported (Chrome 81+).
+ * - If not iframed: all features work properly in both polyfill and built-in.
+ *
+ * @param {function(!Array<!IntersectionObserverEntry>)} ioCallback
+ * @param {!Window} win
+ * @param {string=} threshold
+ *
+ * @return {!IntersectionObserver}
+ */
+export function getViewportObserver(ioCallback, win, threshold) {
+  const iframed = isIframed(win);
+  const root = /** @type {?Element} */ (iframed
+    ? /** @type {*} */ (win.document)
+    : null);
+
+  // TODO(amphtml): Determine if rootMargin is still necessary.
+  const rootMargin = '25%';
+
+  try {
+    return new win.IntersectionObserver(ioCallback, {
+      root,
+      rootMargin,
+      threshold,
+    });
+  } catch (e) {
+    return new win.IntersectionObserver(ioCallback, {rootMargin, threshold});
+  }
+}
+
+/** @type {Map<!Window, !IntersectionObserver>} */
+const viewportObservers = new Map();
+
+/** @type {WeakMap<!Element, function(bool)>} */
+const viewportCallbacks = new WeakMap();
+
+/**
+ * Lazily creates an IntersectionObserver per Window to track when elements
+ * enter and exit the viewport. Fires viewportCallback when this happens.
+ *
+ * @param {!Element} element
+ * @param {function(boolean)} viewportCallback
+ */
+export function observe(element, viewportCallback) {
+  const win = element.ownerDocument.defaultView;
+  if (!viewportObservers.has(win)) {
+    viewportObservers.set(win, getViewportObserver(ioCallback, win));
+  }
+  viewportCallbacks.set(element, viewportCallback);
```

Comment:
Ideally we should do `devAssert(!viewportCallbacks.has(element))` since (a) it's never expected, and (b) multiple legit callbacks is a dev error.

Reply:
I was thinking it is actually useful for the case of build-->layout-->unlayout-->layout.
We may want observe to exist within both build and layout, in which case its fine

Reply:
could replace with:
`devAssert(!viewportCallbacks.has(element) || viewportCallbacks.get(element) === newCb)`

Reply:
The `=== newCb` makes sense to me if that's what you'd prefer. It answers the issue very directly: two *different* callbacks can conflict. If it's the same - no problem.

That being said, I'd rather not nest them. If we can always put them to `layout/unlayout` - great. If not, we'll add some form of `unbuild`. But I really rather not have them nested.

Reply:
what do you mean by "nested". In this case, I think the nesting is effectively a noop right? Its idempotent / safe to do

Reply:
If we call `observe` in `build`, we shouldn't be also calling it in `layout`/`unlayout`. And vice versa. I don't really know if we ever actually need it for `build` callback at this time. The only exception I could imagine are ads and if so - let's discuss.

Reply:
Pushed the change! I'll take extra time in some of the followup PRs to try and avoid any observes in build(). It would be nice to avoid that work. It also has the side-benefit of an extra constraint: vpcb will never be called on an unlaid out component (technically there could be a race between IO and unlayout but thats besides the point)